connectors/memory/memory_test.go

```diff
@@ -707,3 +784,93 @@ func TestPanics(t *testing.T) {
 		passCol(dosa.FieldValue(int64(1)), &dosa.Condition{Op: 0, Value: dosa.FieldValue(int64(1))})
 	})
 }
+
+// createTestData populates some test data. The keyGenFunc can either return a constant,
+// which gives you a single partition of data, or some function of the current offset, which
+// will scatter the data across different partition keys
+func createTestData(t *testing.T, sut *Connector, keyGenFunc func(int) string, idcount int) {
+	// insert a bunch of values with V1 timestamps as clustering keys
+	for x := 0; x < idcount; x++ {
+		err := sut.Upsert(context.TODO(), clusteredEi, map[string]dosa.FieldValue{
+			"f1": dosa.FieldValue(keyGenFunc(x)),
+			"c1": dosa.FieldValue(int64(1)),
+			"c6": dosa.FieldValue(int32(x)),
+			"c7": dosa.FieldValue(dosa.UUID(uuid.NewV1().String()))})
+		assert.NoError(t, err)
+	}
+}
+
+func TestRangePager(t *testing.T) {
+	sut := NewConnector()
+	idcount := 5
+	// create test data in one partition "data"
+	createTestData(t, sut, func(_ int) string { return "data" }, idcount)
+	var token string
+	var err error
+	var data []map[string]dosa.FieldValue
+	for x := 0; x < idcount; x++ {
+		data, token, err = sut.Range(context.TODO(), clusteredEi, map[string][]*dosa.Condition{
+			"f1": {{Op: dosa.Eq, Value: dosa.FieldValue("data")}},
+			"c1": {{Op: dosa.Eq, Value: dosa.FieldValue(int64(1))}},
+		}, dosa.All(), token, 1)
+		assert.NoError(t, err)
+		assert.Len(t, data, 1)
+		if x < idcount-1 {
+			assert.NotEmpty(t, token)
+		} else {
+			assert.Empty(t, token)
+		}
+	}
+	// now walk through again, but delete the item returned
+	// (note: we don't have to reset token, because it should be empty now)
+	for x := 0; x < idcount; x++ {
+		data, token, err = sut.Range(context.TODO(), clusteredEi, map[string][]*dosa.Condition{
+			"f1": {{Op: dosa.Eq, Value: dosa.FieldValue("data")}},
+			"c1": {{Op: dosa.Eq, Value: dosa.FieldValue(int64(1))}},
+		}, dosa.All(), token, 1)
+		assert.NoError(t, err)
+		assert.Len(t, data, 1)
+		if x < idcount-1 {
+			assert.NotEmpty(t, token)
+		} else {
+			assert.Empty(t, token)
+		}
+		err = sut.Remove(context.TODO(), clusteredEi, data[0])
+		assert.NoError(t, err)
+	}
+}
+func TestInvalidToken(t *testing.T) {
+	sut := NewConnector()
+
+	// we don't use the token if there's no data that matches, so lets
+	// create one row
+	createTestData(t, sut, func(id int) string {
+		return "data"
+	}, 1)
+
+	token := "this is not a token and not a hot dog"
+	t.Run("testInvalidTokenRange", func(t *testing.T) {
```

Comment:
The other test, for range on scan, is nearly identical and shares the same setup. The error report when you fail a test is really nice too.